### PR TITLE
Extended dynamic loading

### PR DESCRIPTION
**Pull request by simon-shlomo**
Thanks for your contribution to [Apache Tika](https://tika.apache.org/)! Your help is appreciated!

Before opening the pull request, please verify that
* there is an open issue on the [Tika issue tracker](https://issues.apache.org/jira/projects/TIKA) which describes the problem or the improvement. We cannot accept pull requests without an issue because the change wouldn't be listed in the release notes.
* the issue ID (`TIKA-XXXX`)
  - is referenced in the title of the pull request
  - and placed in front of your commit messages surrounded by square brackets (`[TIKA-XXXX] Issue or pull request title`)
* commits are squashed into a single one (or few commits for larger changes)
* Tika is successfully built and unit tests pass by running `mvn clean test`
* there should be no conflicts when merging the pull request branch into the *recent* `main` branch. If there are conflicts, please try to rebase the pull request branch on top of a freshly pulled `main` branch.

We will be able to faster integrate your pull request if these conditions are met. If you have any questions how to fix your problem or about using Tika in general, please sign up for the [Tika mailing list](http://tika.apache.org/mail-lists.html). Thanks!

### CODE REVIEW

File: tika-core/src/test/java/org/apache/tika/TikaIT.java [hunk 1]
```diff
@@ -27,8 +27,8 @@
     public void testToString() {
         String version = new Tika().toString();
         assertNotNull(version);
-        assertTrue(version.matches(
-                "Apache Tika \\d+\\.\\d+\\.\\d+(-(?:ALPHA|BETA))?(?:-SNAPSHOT)?"));
+//        assertTrue(version.matches(
```
Comment (simon-shlomo) [inline tika-core/src/test/java/org/apache/tika/TikaIT.java:30]: I would keep this .. as the issues is not here.. but in Tika().toString() - it only returns "Apache Tika" (to get a version it needs a pom.properties file - but that does not exist

File: tika-core/src/main/java/org/apache/tika/config/TikaActivator.java [hunk 2]
```diff
@@ -49,15 +61,27 @@ public void start(final BundleContext context) throws Exception {
         bundleContext = context;
 
         detectorTracker = new ServiceTracker(context, Detector.class.getName(), this);
+        encodingDetectorTracker = new ServiceTracker(context, EncodingDetector.class.getName(), this);
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/config/TikaActivator.java:64]: Based on what did you order the trackers (declaration, initialization, open and close)?

File: tika-core/src/main/java/org/apache/tika/detect/AutoDetectReader.java [hunk 3]
```diff
@@ -43,7 +41,8 @@
 public class AutoDetectReader extends BufferedReader {
 
     private static final ServiceLoader DEFAULT_LOADER =
-            new ServiceLoader(AutoDetectReader.class.getClassLoader());
+            new ServiceLoader(AutoDetectReader.class.getClassLoader(), Boolean.getBoolean("org.apache.tika.service.error.warn")
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/AutoDetectReader.java]: How is this going to let AutoDetectReader always use the most recent set of EncodingDetectors?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java [hunk 4]
```diff
@@ -92,7 +90,8 @@ public DefaultDetector(MimeTypes types, ServiceLoader loader) {
     }
 
     public DefaultDetector(MimeTypes types, ClassLoader loader) {
-        this(types, new ServiceLoader(loader));
+        this(types, new ServiceLoader(loader, Boolean.getBoolean("org.apache.tika.service.error.warn")
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java]: This usage of ServiceLoader is repeating. What do you think about adding a constructor signature like this:

`ServiceLoader(ClassLoader loader, boolean dynamic)`

or even just

`ServiceLoader(boolean dynamic)`

Without changing the default behavior of using a non-dynamic behavior by default.

Then we could make the OsgiSpiAdapter capable of instantiating the Tika services properly. E.g. by not using the default constructor anymore, but the additional constructor with the boolean.

What do you think?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java [hunk 5]
```diff
@@ -37,17 +39,35 @@
...
 
     public DefaultEncodingDetector(ServiceLoader loader,
                                    Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors) {
-        super(loader.loadServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
+        super(loader.loadStaticServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
+        this.loader = loader;
+    }
+
+    @Override
+    public List<EncodingDetector> getDetectors() {
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: What about the 'excludeEncodingDetectors' list?

File: tika-core/src/main/java/org/apache/tika/metadata/filter/CompositeMetadataFilter.java [hunk 6]
```diff
@@ -18,7 +18,7 @@
 
 import org.apache.tika.exception.TikaException;
 import org.apache.tika.metadata.Metadata;
-
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/metadata/filter/CompositeMetadataFilter.java]: Maybe Tika is using separating import groups. We should keep (and add) empty lines between those groups

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java [hunk 7]
```diff
@@ -241,12 +247,33 @@ private MediaType detect(ZipArchiveEntry zae, ZipArchiveInputStream zis,
...
+    private List<ZipContainerDetector> getDetectors() {
+        List<ZipContainerDetector> zipDetectors = new ArrayList<>();
+        ServiceLoader loader = getServiceLoader();
+        zipDetectors.addAll(loader.loadDynamicServiceProviders(ZipContainerDetector.class));
+        if (staticZipDetectors0 == null) {
+            staticZipDetectors0 = new ArrayList<ZipContainerDetector>();
+            staticZipDetectors0.addAll(loader.loadStaticServiceProviders(ZipContainerDetector.class));
+        }
+        
+        zipDetectors.addAll(staticZipDetectors0);
```
Comment (thofer) [inline tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: Isn't the concept in Tika, that the passed services are the 'static' services?
So here we only need to add the dynamic services?

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java [hunk 8]
```diff
@@ -241,12 +247,33 @@ private MediaType detect(ZipArchiveEntry zae, ZipArchiveInputStream zis,
...
+        return zipDetectors;
+    }
+
+    private ServiceLoader getServiceLoader() {
```
Comment (thofer) [inline tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: I suggest to try to use the same pattern for getting the service loader as in all other classes.

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/StreamingZipContainerDetector.java [hunk 9]
```diff
@@ -85,4 +86,5 @@ public MediaType detect(InputStream input, Metadata metadata) throws IOException
             return detectCompressorFormat(prefix, length);
         }
     }
+
```
Comment (thofer) [inline tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/StreamingZipContainerDetector.java]: Undo unneeded change

**Review by thofer — CHANGES_REQUESTED** [review]
Why do you commit deletion of 'droste.zip' file?

File: tika-core/src/main/java/org/apache/tika/config/TikaActivator.java (hunk 2 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/config/TikaActivator.java:64]: order in the OsgiSpiAdaptor - but you right it would be more readable if it was alphabetical ordered here

File: tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java (hunk 4 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java]: I am reviewing all the changes again. I think the Tika developers want that when you pass a classloader it means *only* this class loader is used. E.g. here in the DefaultDetector the default if you do not pass a classloader is to also load dynamic services - if you pass a class loader it only loads the classes the class loader knows.

So now I do not see the need to change the default behaviour here..

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 7 above)
Comment (simon-shlomo) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: I cannot follow? If we do not add the staticZipDetectors0 here this will not be used. getDetectors() is supposed to return all detectors

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 8 above)
Comment (simon-shlomo) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: okay - just means a bit more complex getDetectors - because this class works differently as the other as it has constructor to directly set the "detectors" (but this could be null), and if they are set the loader is null

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 7 above)
Comment (simon-shlomo) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: > Why do you commit deletion of 'droste.zip' file?
I will add it back as the last commit - it complex to keep track of this external deletion.

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java (hunk 5 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: In the parent class the exclude list is only used for the static services - therefore, I didn't add an exclusion here - but it does makes sense that the exclusion list is used - now added

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java [hunk 10]
```diff
@@ -71,10 +79,15 @@ public Charset detect(InputStream input, Metadata metadata) throws IOException {
...
+        List<EncodingDetector> allDetectors = new LinkedList<>();
+        if (loader != null) {
+            allDetectors.addAll(loader.loadDynamicServiceProviders(EncodingDetector.class));
+        }
+        allDetectors.addAll(detectors);
+        return allDetectors;
     }
 
-    private boolean isExcluded(Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors,
+    protected boolean isExcluded(Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors,
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: Why does this need to be protected? Shouldn't 'getDetectors' use the 'isExcluded' method?

File: tika-core/src/main/java/org/apache/tika/detect/AutoDetectReader.java [hunk 11]
```diff
@@ -43,14 +43,7 @@
...
     private static final ServiceLoader DEFAULT_LOADER =
-            new ServiceLoader(AutoDetectReader.class.getClassLoader());
-
-    private static EncodingDetector DEFAULT_DETECTOR;
-
-    static {
-        DEFAULT_DETECTOR = new CompositeEncodingDetector(
-                DEFAULT_LOADER.loadServiceProviders(EncodingDetector.class));
-    }
+            new ServiceLoader(AutoDetectReader.class.getClassLoader(), true);
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/AutoDetectReader.java]: AutoDetectReader is a special class. How is it used? What constructors are really used?

Shouldn't we try to get the 'default' behavior of TIKA to be the same as before, but enabling 'dynamic' behaviour for our case? Is that doable?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java [hunk 12]
```diff
@@ -17,11 +17,8 @@
 package org.apache.tika.detect;
 
 import java.util.Collection;
-import java.util.Comparator;
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java]: Why do we have these changes?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java [hunk 13]
```diff
@@ -37,17 +38,44 @@
...
                                    Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors) {
-        super(loader.loadServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
+        super(loader.loadStaticServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
+        this.excludeEncodingDetectors = excludeEncodingDetectors;
+        this.loader = loader;
+    }
+
+    @Override
+    public List<EncodingDetector> getDetectors() {
+        if (loader != null) {
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: What is the difference of 'getDetectors' here and in the parent class?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java [hunk 14]
```diff
@@ -37,17 +38,44 @@
...
 
+    private static final long serialVersionUID = 1L;
+    
+    private ServiceLoader loader;
+    private Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors;
+
+
     public DefaultEncodingDetector() {
-        this(new ServiceLoader(DefaultEncodingDetector.class.getClassLoader()));
+        this(new ServiceLoader(DefaultEncodingDetector.class.getClassLoader(), true));
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: In my last review I came up with the idea that we should try to keep old behavior as-is but adding new behavior using the additional 'dynamic' entry point.

I assume that such a change is possible, if we (as I wrote) introduce a new constructor in all services which our OsgiSpiAdapter is calling (or something similar).

**Review by thofer — CHANGES_REQUESTED** [review]

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 10 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: the getDetector of the subclass also need isExcluded

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java (hunk 13 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: it excludes the detectors that are on the list of excluded detector.. cf. your suggestion..

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java [hunk 15]
```diff
@@ -71,10 +80,15 @@ public Charset detect(InputStream input, Metadata metadata) throws IOException {
     }
 
     public List<EncodingDetector> getDetectors() {
-        return Collections.unmodifiableList(detectors);
+        List<EncodingDetector> allDetectors = new LinkedList<>();
+        if (loader != null) {
+            allDetectors.addAll(loader.loadDynamicServiceProviders(EncodingDetector.class));
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: Why don't you filter the dynamic loaded services?

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 10 above)
Comment (thofer) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: Why does the subclass need to filter?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java [hunk 16]
```diff
@@ -18,6 +18,7 @@
 
 import java.util.Collection;
 import java.util.Comparator;
+
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultDetector.java]: Is this change needed?

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java (hunk 13 above)
Comment (thofer) [reply tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: But it's the parent class that introduces the 'excludeEncodingDetectors' feature, not this one.

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java [hunk 17]
```diff
@@ -241,12 +251,27 @@ private MediaType detect(ZipArchiveEntry zae, ZipArchiveInputStream zis,
...
             }
         }
         return MediaType.APPLICATION_ZIP;
     }
+
+    private List<ZipContainerDetector> getDetectors() {
+        List<ZipContainerDetector> zipDetectors = new ArrayList<>();
+        if (loader != null) {
+            zipDetectors.addAll(loader.loadDynamicServiceProviders(ZipContainerDetector.class));
+            if (staticZipDetectors == null) {
```
Comment (thofer) [inline tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: Why does this class handle the 'static' services differently than the other classes?

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/StreamingZipContainerDetector.java [hunk 18]
```diff
@@ -38,19 +39,19 @@
...
-    List<ZipContainerDetector> zipDetectors;
+    
 
     public StreamingZipContainerDetector() {
-        this(new ServiceLoader(StreamingZipContainerDetector.class.getClassLoader()));
+       // Empty
     }
 
     public StreamingZipContainerDetector(ServiceLoader loader) {
-        this(loader.loadServiceProviders(ZipContainerDetector.class));
```
Comment (thofer) [inline tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/StreamingZipContainerDetector.java:48]: Is there no need for the 'dynamic' constructor?

**Review by thofer — CHANGES_REQUESTED** [review]

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 15 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: that part I now removed to make it similar to the structure of the DefatultDetector

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 10 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: becuase the subclass loads the dynamic services - this class only knows about the static (this is similar to how the defaultDetector and CopositeDetector works)

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java (hunk 13 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: I am trying to have similar structure as the DefaultDetector so dynamic services are only loaded in the subclass

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 17 above)
Comment (simon-shlomo) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: because the other classes have a superclass that returns the static detectors in getDetectors() but this class does not, so it needs to store them in staticZipDetectors

Comment (thofer) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: Do you have other examples where the static service providers are loaded lazily? I thought that the static providers get loaded eagerly everywhere else.

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 10 above)
Comment (thofer) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: But in CompositeDetector the isExcluded method is private too...

Implementation of CompositeDetector is bad, because both subclasses have to implement loading the dynamic services by their own. That's against the idea of inheritance.

In addition to that the sub classes of CompositeDetector do not filter based on excluded list. So then we go the simple path and do it the same way: we don't filter the dynamic.

How many implementations with CompositeEncodingDetector as parent class do exist?

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 17 above)
Comment (thofer) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: Do you have other examples where the static service providers are loaded lazily? I thought that the static providers get loaded eagerly everywhere else.

**Review by thofer — CHANGES_REQUESTED** [review]

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java (hunk 10 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: Okay then I will remove the filtering of the dynamic

CompositeEncodingDetector has one subclass (that I can find by searching on github) - i.e. the DefaultEncodingDetector

File: tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java (hunk 17 above)
Comment (simon-shlomo) [reply tika-parsers/tika-parsers-classic/tika-parsers-classic-modules/tika-parser-zip-commons/src/main/java/org/apache/tika/detect/zip/DefaultZipContainerDetector.java]: No I don't - I will move it to the constructor if it problematic to load lazily

File: tika-core/src/main/java/org/apache/tika/config/TikaConfig.java [hunk 19]
```diff
@@ -681,7 +681,7 @@ T loadOne(Element element, MimeTypes mimeTypes, ServiceLoader loader)
                     // Regular class, create as-is
                     loaded = newInstance(loadedClass);
                     // TODO Support arguments, needed for Translators etc
-                    // See the thread "Configuring parsers and translators" for details 
+                    // See the thread "Configuring parsers and staticTranslators" for details
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/config/TikaConfig.java]: Undo

File: tika-core/src/main/java/org/apache/tika/config/TikaConfigSerializer.java [hunk 20]
```diff
@@ -129,7 +129,7 @@ private static void addServiceLoader(Mode mode, Element rootElement, Document do
 
     private static void addTranslator(Mode mode, Element rootElement, Document doc, TikaConfig config) {
         // Unlike the other entries, TikaConfig only wants one of
-        //  these, and no outer <translators> list
+        //  these, and no outer <staticTranslators> list
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/config/TikaConfigSerializer.java]: Undo

File: tika-core/src/main/java/org/apache/tika/config/TikaConfigSerializer.java [hunk 21]
```diff
@@ -145,7 +145,7 @@ private static void addTranslator(Mode mode, Element rootElement, Document doc,
                 translatorElement.setAttribute("class", translator.getClass().getCanonicalName());
                 rootElement.appendChild(translatorElement);
             } else {
-                rootElement.appendChild(doc.createComment("No translators available"));
+                rootElement.appendChild(doc.createComment("No staticTranslators available"));
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/config/TikaConfigSerializer.java]: Undo

File: tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java [hunk 22]
```diff
@@ -37,17 +40,43 @@
...
-        super(loader.loadServiceProviders(EncodingDetector.class));
+        super(loader.loadStaticServiceProviders(EncodingDetector.class));
+        dynamicDetectors = loader.loadDynamicServiceProviders(EncodingDetector.class);
+        this.loader = loader;
     }
 
     public DefaultEncodingDetector(ServiceLoader loader,
                                    Collection<Class<? extends EncodingDetector>> excludeEncodingDetectors) {
-        super(loader.loadServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
+        super(loader.loadStaticServiceProviders(EncodingDetector.class), excludeEncodingDetectors);
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/DefaultEncodingDetector.java]: This is going to change the old behavior. The old behavior might have loaded static and dynamic providers. Now only static providers are loaded and filtered.

File: tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java [hunk 23]
```diff
@@ -52,6 +53,11 @@ public CompositeEncodingDetector(List<EncodingDetector> detectors) {
         this.detectors.addAll(detectors);
     }
 
+    public CompositeEncodingDetector(ServiceLoader loader) {
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/detect/CompositeEncodingDetector.java]: Is this constructor needed?

File: tika-core/src/main/java/org/apache/tika/extractor/DefaultEmbeddedStreamTranslator.java [hunk 24]
```diff
@@ -42,11 +47,23 @@
...
+    }
+    
+    public DefaultEmbeddedStreamTranslator(boolean loadDynamicServices) {
+        this(new ServiceLoader(loadDynamicServices));
+        this.loadDynamicServices = loadDynamicServices;
+    }
+    
+    private DefaultEmbeddedStreamTranslator(ServiceLoader loader) {
+        this(getDefaultFilters(loader));
+        dynamicTranslators = loader.loadDynamicServiceProviders(EmbeddedStreamTranslator.class);
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/extractor/DefaultEmbeddedStreamTranslator.java]: I think that in case of 'loadDynamicServices' the 'dynamic translators' end up two times in the resulting list.

File: tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java [hunk 25]
```diff
@@ -42,11 +42,11 @@ public DefaultTranslator() {
     }
 
     /**
-     * Finds all statically loadable translators and sort the list by name,
+     * Finds all statically loadable staticTranslators and sort the list by name,
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java]: Undo

File: tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java [hunk 26]
```diff
@@ -42,11 +42,11 @@ public DefaultTranslator() {
...
      * rather than discovery order.
      *
      * @param loader service loader
-     * @return ordered list of statically loadable translators
+     * @return ordered list of statically loadable staticTranslators
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java]: Undo

File: tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java [hunk 27]
```diff
@@ -71,7 +71,7 @@ public String translate(String text, String sourceLanguage, String targetLanguag
         if (t != null) {
             return t.translate(text, sourceLanguage, targetLanguage);
         }
-        throw new TikaException("No translators currently available");
+        throw new TikaException("No staticTranslators currently available");
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java]: Undo

File: tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java [hunk 28]
```diff
@@ -82,11 +82,11 @@ public String translate(String text, String targetLanguage) throws TikaException
         if (t != null) {
             return t.translate(text, targetLanguage);
         }
-        throw new TikaException("No translators currently available");
+        throw new TikaException("No staticTranslators currently available");
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java]: Undo

File: tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java [hunk 29]
```diff
@@ -82,11 +82,11 @@ public String translate(String text, String targetLanguage) throws TikaException
...
     }
     
     /**
-     * Returns all available translators
+     * Returns all available staticTranslators
```
Comment (thofer) [inline tika-core/src/main/java/org/apache/tika/language/translate/DefaultTranslator.java]: Undo

**Review by thofer — CHANGES_REQUESTED** [review]

File: tika-core/src/main/java/org/apache/tika/config/TikaConfig.java (hunk 19 above)
Comment (simon-shlomo) [reply tika-core/src/main/java/org/apache/tika/config/TikaConfig.java]: funny this sneaked in.. eclipse has its own life

File: tika-core/src/main/java/org/apache/tika/config/TikaActivator.java [hunk 30]
```diff
@@ -40,24 +43,44 @@
 
     private ServiceTracker detectorTracker;
 
+    private ServiceTracker embeddedStreamTranslatorTracker;
```
Comment (simon-shlomo) [inline tika-core/src/main/java/org/apache/tika/config/TikaActivator.java:46]: These might not be needed - but - I was checking few of the parser packages and found a package providing an EncodingDetector - so I suggest we keep them to avoid issues with providers not being loaded.

**Review by thofer — APPROVED** [review]
